tests/integration/ha/test_network_cut.py

```diff
@@ -0,0 +1,381 @@
+    )
+
+    logger.info(f"Waiting until unit {old_leader_name} is 'lost'")
+    await ops_test.model.block_until(
+        lambda: ["unknown", "lost"]
+        == ha_helpers.get_unit_state_from_status(ops_test, old_leader_name),
+        timeout=LONG_TIMEOUT,
+        wait_period=LONG_WAIT,
+    )
+
+    logger.info("Checking leader re-election...")
+    new_leader_name = await get_leader_name(ops_test)
+    assert new_leader_name != old_leader_name
+
+    logger.info("Checking Dashboard access for the rest of the nodes...")
+    assert await access_all_dashboards(ops_test, skip=[old_leader_name], https=https)
+
+    logger.info("Restoring network...")
+    try:
+        ha_helpers.network_release(machine_name)
+    except CalledProcessError:  # in case it was already cleaned up
+        pass
+
+    logger.info(f"Waiting until unit {old_leader_name} is reachable again")
+    await ops_test.model.block_until(
+        lambda: ha_helpers.reachable(old_ip, SERVER_PORT),
+        timeout=LONG_TIMEOUT,
+        wait_period=LONG_WAIT,
+    )
+
+    logger.info("Checking Dashboard access...")
+    assert await access_all_dashboards(ops_test, https=https)
+
+
+async def network_cut_application(ops_test: OpsTest, https: bool = False):
+    """Full network cut for the whole application, resulting in IP change."""
+    logger.info("Cutting all units from network...")
+
+    machines = []
+    units = []
+    ips = []
+    for unit in ops_test.model.applications[APP_NAME].units:
+        machine_name = await ha_helpers.get_unit_machine_name(ops_test, unit.name)
+        ip = await get_address(ops_test, unit.name)
+
+        logger.info("Cutting unit {unit.name} from network...")
+        ha_helpers.cut_unit_network(machine_name)
+
+        machines.append(machine_name)
+        units.append(unit.name)
+        ips.append(ip)
+
+    logger.info(f"Waiting until units {units} are not reachable")
+    await ops_test.model.block_until(
+        lambda: not all(ha_helpers.reachable(ip, SERVER_PORT) for ip in ips),
+        timeout=LONG_TIMEOUT,
+        wait_period=LONG_WAIT,
+    )
+
+    logger.info(f"Waiting until unit {units} are 'lost'")
+    await ops_test.model.block_until(
+        lambda: all(
+            ["unknown", "lost"] == ha_helpers.get_unit_state_from_status(ops_test, unit)
+            for unit in units
+        ),
+        timeout=LONG_TIMEOUT,
+        wait_period=LONG_WAIT,
+    )
+
+    logger.info("Checking lack of Dashboard access...")
+    assert not (await access_all_dashboards(ops_test, https=https))
+
+    logger.info("Restoring network...")
+    for machine_name in machines:
+        try:
+            ha_helpers.restore_unit_network(machine_name)
+        except CalledProcessError:  # in case it was already cleaned up
+            pass
+
+    logger.info("Waiting for Juju to detect new IPs...")
+    await ops_test.model.block_until(
+        lambda: not any(ip in ha_helpers.get_hosts_from_status(ops_test).values() for ip in ips),
```

Comment:
This is risky, I don't think we have the guarantee that a previously allocated IP will not be reused for **another** unit. 
We should rather test the change of IPs for each unit `unit1_old_ip VS unit1_new_ip`

Reply:
The same check is considered safe on other Data Platform charms (perhaps for a low likelyhood on regularly running pipelines). Neverthless we can make it even more secure.

Reply:
Changed as requested.